### ISSUE TITLE
Inverse reward metrics

### ISSUE THREAD

**Issue by m-ronchi** [Unity-Technologies/ml-agents#846]
Hi,
I am training 2 agents in a 1vs1 game. (card game)
I setup the rewards as follows:
- -0.5 for making an invalid action (i.e. playing a card not in your hand) (then the game is reset)
- +0.05 for making a valid action (game has fixed 20 actions per player, 40 total)
- +1 to the winner
- +1/240 for each point taken (game has a fixed total score of 120)

so, the total reward of a successful game, assuming no invalid actions taken, is `(0.05*40 + 1 + 120/240)` = `2.5`.
therefore the cumulative reward is capped at 1.75.
(it may be less for the following reasons:
- draws are allowed (but unlikely to happen randomly). in case of a draw, I don't give the +1 reward to anyone
- due to #842, I abort the game when the agent makes 3 invalid actions in a row (playing randomly the first 2 times, meaning an occasional punishment can still happen

)
this value is confirmed, as I hit a cap of ~1.66 for the `cumulative_reward` stat in tensorboard.

main question is: which metric should I observe in TB to determine if my agents are still learning?

**Comment by vincentpierre** [comment]
Hi, I was unable to reproduce your calculations for the cumulative reward. Also, you seem to be making a distinction between total reward and cumulative reward. I do not understand: The agent gets one reward per step, the cumulative reward is the sum of all rewards obtained during an episode (between two agent reset).
Regarding which metric to look at, this is an interesting problem. I would say testing the Neural Network against a past version of itself would be a good idea. I would say look at your Entropy metric, it should have decreased and stabilized to a non-zero value.

**Comment by m-ronchi** [comment]
Hi,
by total reward, I mean the sum of all cumulative rewards of all agent each game (episode). that is constant, as my game features:
- 40 actions total (20 per agent) -> each gets +1 cumulative reward
- "almost" exactly one winner (game can be drawn) -> +1 reward should go to one agent
- 120 points total -> +0.5 reward is spread between agents

therefore, the "reward pool" for each episode is fixed and distributed between both players, assuming
- game not drawn -> I may want to change this by giving +0.5 to each agent if the game is drawn, but I also want them to consider a draw the same as a loss. what should the best strategy be here?
- no invalid actions made -> this would be the first thing the agents learned, as I observed the graph rise from less than -1 (and short episode length) to that 1.66 value (and episode length of 19)
the average cumulative reward that I see in tensorboard should be half of that total.

ps: Actually, I just noticed that the reward pool should be 3.5, not 2.5, so the value I am seeing is off.
I just wanted to confirm my reasoning:
for a fixed pool / inverse reward functions, the TB value should be `total_reward / number of agents` right?

**Comment by vincentpierre** [comment]
If your two agents are matched there should be a lot of draws I would assume. Maybe you could log other metrics in Unity if you want to know if your agent makes invalid moves. Training a Neural Network for a zero sum games against itself can cause problems because the agent has no incentive to get better (since the opponent would get better as well). I would try to train two external brains against each other (This can cause other issues as one agent could get stuck, but I think it should be better in a zero sum game context). Another option would be to train it against some Heuristic or a pre-trained network. If you do this then you will be able to see progress more effectively.
I hope this helps.

**Comment by eshvk** [comment]
Thanks for reaching out to us. Hopefully you were able to resolve your issue.  We are closing this due to inactivity, but if you need additional assistance, feel free to reopen the issue.
